src/const_eval.md

```diff
@@ -23,7 +23,8 @@ to be run.
 * [Const parameters].
 * [Paths] to [functions] and [constants].
   Recursively defining constants is not allowed.
-* Paths to [statics]. These are only allowed within the initializer of a static.
+* Paths to immutable [statics].
+    * Reads of [`extern` statics] are not allowed.
```

Comment:
Reads from and writes to `static mut` and `static` with interior mutability are also not allowed.

And then there are restrictions on references that escape into the final value of the const/static. More precisely, this is about expressions that have been subject to lifetime extension, also see https://github.com/rust-lang/reference/pull/1637. Those references must also be of type `&T` and point to a value without an `UnsafeCell`.

And finally, specifically for `const` there also can't be any references to anything mutable in the final value.

Reply:
Hm, I'm trying to unpack what you are saying here and align that with the current rules.

> Reads from and writes to `static mut` …

Isn't this implied by the fact that you are not allowed to refer to `static mut` at all by the rule that says paths only to *immutable* statics are allowed?

> Reads from and writes to [..] `static` with interior mutability are also not allowed.

Thanks, I have added that.

> And then there are restrictions on references that escape into the final value of the const/static.

Isn't this specified in the later rule that limits borrows of `UnsafeCell` that are transient? (That was my reading of it, and why I didn't say it explicitly.)

> And finally, specifically for `const` there also can't be any references to anything mutable in the final value.

Thanks, I have added that to the constant-items chapter.

Reply:
I actually have found the points mentioned down in the list here, in a bullet point starting with `All forms of [borrow]s, including raw borrows, with one limitation ...`. I think we have it covered.

Reply:
> Isn't this specified in the later rule that limits borrows of UnsafeCell that are transient? (That was my reading of it, and why I didn't say it explicitly.)

Ah right, I forgot that those are already documented.

Reply:
> That's not correct, referring to static mut is permitted.

OK, then I go back to my comment [earlier](https://github.com/rust-lang/reference/pull/1610#discussion_r1779733601) about why the word "immutable" is being added here.

I would think this PR shouldn't add the word *immutable*, and the edit @dingxiangfei2009 just added about not being allowed to read or write from `static mut` should cover everything?

(Because as written now, it says you can only do paths to immutable statics, but that does not seem to be true based on the playground above.)

Reply:
Yeah you are right, `immutable` should not be mentioned here.